python_utils_test.py

```diff
@@ -205,13 +205,21 @@ def test_recursively_convert_to_str_with_list(self):
         for item in list_in_str:
             self.assertNotEqual(type(item), future.types.newstr)
             self.assertNotEqual(type(item), future.types.newbytes)
+            self.assertTrue(isinstance(item, dict))
```

Comment:
OK, I'm glad we did this. It exposes a problem with this test.

There is no point asserting that a dict is not a future.types.newstr. Of course it won't be. The question is more one of newstr/newbytes vs str/unicode.

So, instead, please update these checks to be more robust by ensuring that all your checks for "not str/bytes" are made on things that are at least some kind of string-like thing.

Reply:
Actually the list contained only one element, so line 205-208 was not checking anything. Its the checks after this block that counts. So I changed the list to be a dict itself

Reply:
My comment still stands though. The things that you are checking are not equal to newstr or newbytes should fulfill an assertTrue check of "it is a str / unicode / bytes" (as appropriate). You still have checks against lists or dicts, which are trivially not strings.

Reply:
I am now testing all the elements and I have checks for unicode/bytes too. For example, since the key is a list, I am also iterating through it and testing its values too

Reply:
What I'm trying to say is -- I don't see why you need these statements:

https://github.com/oppia/oppia/pull/7446/files#diff-7fe513e42480cc4544432522bf6e88a1R208
https://github.com/oppia/oppia/pull/7446/files#diff-7fe513e42480cc4544432522bf6e88a1R209

because "value" is clearly a list.

Also, this statement: https://github.com/oppia/oppia/pull/7446/files#diff-7fe513e42480cc4544432522bf6e88a1R216 should not have a dict in it. (Which means you may need to unwrap it and test the strings inside it explicitly.)

Reply:
Done